src/Mutator/Util/AbstractSingleSubExprNegation.php

```diff
@@ -0,0 +1,128 @@
+ *
+ * * Redistributions of source code must retain the above copyright notice, this
+ *   list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of the copyright holder nor the names of its
+ *   contributors may be used to endorse or promote products derived from
+ *   this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+declare(strict_types=1);
+
+namespace Infection\Mutator\Util;
+
+use Infection\Mutator\Mutator;
+use Infection\Mutator\SimpleExpression;
+use Infection\PhpParser\Visitor\ParentConnector;
+use LogicException;
+use PhpParser\Node;
+
+/**
+ * @internal
+ *
+ * @template TNode of Node
+ * @implements Mutator<TNode>
+ */
+abstract class AbstractSingleSubExprNegation implements Mutator
+{
+    use SimpleExpression;
+
+    /**
+     * @psalm-mutation-free
+     *
+     * @param Node\Expr $node
+     */
+    public function mutate(Node $node): iterable
+    {
+        $subExpressionsToNegateCount = $this->countSubExpressionsToNegate($node);
+
+        for ($expressionIndex = 0; $expressionIndex < $subExpressionsToNegateCount; ++$expressionIndex) {
+            yield $this->negateSubExpression($node, $expressionIndex);
+        }
+    }
+
+    public function canMutate(Node $node): bool
+    {
+        if (!$this->isInstanceOf($node)) {
+            return false;
+        }
+
+        $parent = ParentConnector::findParent($node);
+
+        return $parent !== null && !$this->isInstanceOf($parent); // only grandparent
+    }
+
+    abstract protected function isInstanceOf(Node $node): bool;
+
+    /**
+     * @param array<string, mixed> $attributes
+     */
+    abstract protected function create(Node\Expr $left, Node\Expr $right, array $attributes): Node\Expr;
+
+    private function countSubExpressionsToNegate(Node\Expr $node, int &$count = 0): int
+    {
+        if ($this->isInstanceOf($node)) {
+            if (!isset($node->left, $node->right)) {
```

Comment:
from what I can see only `BinaryOp` nodes can have left & right within the `Expr` nodes, so maybe:

- `isInstanceOf` could be replaced by `isSupportedBinaryOpExpr(Node\Expr\AssignOp): bool`
- the existing `isInstanceOf()` calls could be combined with a `Node\Expr\BinaryOp` `instanceof` check – or this could be abstracted into a `isSupportedNode()` private method

This would also get rid of this `LogicException` thrown